File: src/endpoints/checklist/get_item.py
from src import output
from src.libs.db import DB, Table
from src.modules.verify import checking_token
from src.modules.mod import MOD
from . import __types__ as types
from ..tag import __libs__ as tag_libs

async def get_item(params: dict = {}, req = None, _db: DB = None, _token = None):

    # set values
    result = None
    db = _db if _db else DB().connect()

    try:
        # set params
        params = types.GetItem(**params)

        # checking token
        token = checking_token(req, db) if not _token else _token

        # set fields
        fields = params.fields.split(',') if params.fields else None

        # set mod
        mod = MOD(params.mod or '')

        # get data
        data = db.get_item(
            table_name=Table.CHECKLIST.value,
            fields=fields,
            where=[ f'srl = {params.srl}' ],
        )
        if not data: raise Exception('Item not found', 204)

        # MOD / count-file
        if mod.check('count-file'):
            count = db.get_count(
                table_name=Table.FILE.value,
                where=[
                    f'AND module=\'checklist\'',
                    f'AND module_srl = {params.srl}',
                ]
            )
            data['count_file'] = count or 0

        # MOD / tag
        if mod.check('tag'):
            _tags = tag_libs.get_index(
                _db=db,
                module=tag_libs.Module.CHECKLIST,
                module_srl=params.srl,
            )
            data['tag'] = _tags if _tags and len(_tags) > 0 else None

        # set result
        result = output.success({
            'message': 'Complete get checklist item.',
            'data': data,
        }, _req=req)
    except Exception as e:
        result = output.exc(e, _req=req)
    finally:
        if not _db and db: db.disconnect()
        return result

File: src/routes/checklist/get-item.ts
import DB, { db } from '@/classes/DB'
import ServiceError from '@/classes/ServiceError'
import MOD from '@/classes/MOD'
import * as fileHelper from '@/routes/file/__helper'
import * as tagHelper from '@/routes/tag/__helper'
import type { ChecklistModel } from './__model'

type GetItemParams = {
  srl: number
  query: ChecklistModel['getItemQuery']
}

export default async function getItem({ srl, query }: GetItemParams)
{
  try
  {
    // set field
    const _field = query.field ? query.field.split(',') : ''

    // get data
    const item = db.getData({
      table: DB.TABLE.CHECKLIST,
      field: _field,
      where: `srl = ${srl}`,
    })
    if (!item.data) throw new ServiceError('No data', { status: 204 })

    // set mod
    const _mod: MOD = new MOD(query.mod)

    // MOD / count-file
    if (_mod.check('count-file'))
    {
      item.data.count_file = fileHelper.count({
        where: [
          `AND module LIKE \'${fileHelper.MODULE.CHECKLIST}\'`,
          `AND module_srl = ${srl}`,
        ],
      })
    }
    // MOD / tag
    if (_mod.check('tag'))
    {
      item.data.tag = tagHelper.getIndex(tagHelper.MODULE.CHECKLIST, srl)
    }

    // return
    return {
      ...item.data,
    }
  }
  catch (_e: any)
  {
    throw new ServiceError('Failed to get Checklist.', {
      status: _e.status,
      text: _e.message,
      cause: _e,
    })
  }
}


